examples/todo/js/NetworkLayer.js

```diff
@@ -0,0 +1,78 @@
+/* eslint-disable no-console */
+
+import invariant from 'invariant';
+import Relay from 'react-relay';
+import io from 'socket.io-client';
+
+export default class NetworkLayer extends Relay.DefaultNetworkLayer {
+  constructor(...args) {
+    super(...args);
+
+    this._socket = io();
+    this._observers = {};
+
+    this._socket.on('subscription update', ({ id, data, errors }) => {
+      const observer = this._observers[id];
+      if (!observer) return;
+
+      if (errors) {
+        observer.onError(errors);
+      } else {
+        observer.onNext(data);
+      }
+    });
+
+    this._socket.on('subscription closed', (id) => {
+      const observer = this._observers[id];
+      if (!observer) {
+        return;
+      }
+
+      console.log(`Subscription ${id} is completed`);
+      observer.onCompleted();
+      delete this._observers[id];
+    });
+
+    this._socket.on('error', (error) => {
+      Object.values(this._observers).forEach((observer) => {
+        observer.onError(error);
+      });
+    });
+  }
+
+  sendSubscription(request) {
+    const id = request.getClientSubscriptionId();
+
+    return {
+      subscribe: (observer) => {
+        invariant(
+          !this._observers[id],
```

Comment:
This check will never fail in practice, since `subscribe` only gets called once and there's no other way to access this observable.